controllers/chat/getSignedChannelList.js

```diff
@@ -0,0 +1,73 @@
+const {responseSuccess} = require('../../utils/Responses');
+const GetstreamSingleton = require('../../vendor/getstream/singleton');
+
+const __queryChannelBuilder = async (userId, limit, offset) => {
+  console.log('querying channel', limit, offset);
+  const client = GetstreamSingleton.getChatInstance();
+  const channels = await client.queryChannels({members: {$in: [userId]}}, [{last_message_at: -1}], {
+    limit: limit,
+    offset: offset,
+    state: true
+  });
+  await client.disconnectUser();
+  return channels;
+};
+
+const getSignedChannelList = async (req, res) => {
+  const {userId} = req;
+  const {limit = 100, offset = 0} = req.query;
+
+  let totalFetched = 0;
+  const queriedChannels = [];
+  const promisedChannels = [];
+
+  const client = GetstreamSingleton.getChatInstance();
+  try {
+    await client.connectUser({id: req.userId}, req.token);
+
+    while (totalFetched < limit) {
+      const batchSize = Math.min(30, limit - totalFetched);
+      promisedChannels.push(
+        __queryChannelBuilder(userId, batchSize, Number(totalFetched) + Number(offset))
+      );
+      totalFetched += batchSize;
+      if (totalFetched >= limit) {
+        break;
+      }
+    }
+
+    const responses = await Promise.all(promisedChannels);
+    responses.forEach((response) => {
+      queriedChannels.push(...response);
+    });
+
+    await client.disconnectUser();
+    const channels = queriedChannels.map((channel) => {
+      const newChannel = {...channel.data};
+      delete newChannel.config;
+      delete newChannel.own_capabilities;
+
+      const members = [];
+      Object.keys(channel.state.members).forEach((member) => {
+        members.push(channel?.state?.members[member]);
+      });
+
+      return {
+        ...newChannel,
+        members,
+        unreadCount: channel.state.unreadCount,
+        messages: channel.state.messages
+      };
+    });
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The mapping function used to transform the queried channels is quite complex and could be simplified. Consider extracting it into a separate function for better readability and maintainability.

```diff
-    const channels = queriedChannels.map((channel) => {
-      const newChannel = {...channel.data};
-      delete newChannel.config;
-      delete newChannel.own_capabilities;
-
-      const members = [];
-      Object.keys(channel.state.members).forEach((member) => {
-        members.push(channel?.state?.members[member]);
-      });
-
-      return {
-        ...newChannel,
-        members,
-        unreadCount: channel.state.unreadCount,
-        messages: channel.state.messages
-      };
-    });
+    const channels = queriedChannels.map(transformChannelData);
```

Where `transformChannelData` is a new function that encapsulates the transformation logic:

```javascript
function transformChannelData(channel) {
  const newChannel = {...channel.data};
  delete newChannel.config;
  delete newChannel.own_capabilities;

  const members = Object.values(channel?.state?.members || {});

  return {
    ...newChannel,
    members,
    unreadCount: channel.state.unreadCount,
    messages: channel.state.messages
  };
}
```

This makes the code more modular and easier to test and maintain.


<!-- This is an auto-generated comment by OSS CodeRabbit -->